### PR TITLE
Modifying the optimiser input to more elaborate `PREDICTION` vector dtype

### LINKED ISSUE #9: Set edge and node probabilities to be used by the ILP optimiser
Once #7 and #8 are implemented, we can set up the optimiser. This requires:

+ [ ] `GraphAttrs.NODE_PROB_DETECTION` to be a feature of the networkx nodes
+ [ ] `GraphAttrs.EDGE_PROB_LINK` to be a feature of the networkx edges
+ [ ] addition of some scaling constant

see:
https://github.com/alan-turing-institute/grace/blob/ad2a3d86668d2e457cd26688ae82ba318169c1b7/grace/models/optimise.py#L141-L153

A current test implementation is in `opt.ipynb`

### LINKED ISSUE #191: Modify optimiser to read more elaborate `GraphAttrs.PREDICTION` data structures
Currently, optimiser reads the node / edge predictions as simple `float` numbers:

https://github.com/alan-turing-institute/grace/blob/e634d3621747d975f974aade96141bb9174d5594/grace/models/optimiser.py#L141-L153

This needs to be updated for more elaborate prediction storage...

### LINKED ISSUE #213: Repurpose `GraphAttrs.Prediction` attributes to complete set
Currently, the (dummy) classifier & optimiser rely on the predictions to be a single `float`:

https://github.com/alan-turing-institute/grace/blob/7ead56fdbbb0c0a061d35a36899bd5115cd4ab09/grace/base.py#L103-L105

This is incomplete and needs to be addressed to store a `tuple` where `(0, [0.88, 0.12])` where the [0]-th item is the `argmax` index of the predicted class, and the [1]-st item is the list of `softmax` probabilities per respective class.

### LINKED ISSUE #226: [DEVELOPMENT] Flip the edge probability multiplication from 0.1 to 0.9
This will help with experimentation of quantification of how certain (yet always correct) the optimiser has to be for optimal object detection:

https://github.com/alan-turing-institute/grace/blob/488f60bf6822106129c1de1c83b244c051dfc827/grace/evaluation/process.py#L7-L11

### LINKED ISSUE #228: [DEVELOPMENT] Edit optimiser hypothesis rho value input type

### PR DESCRIPTION
**Pull request by KristinaUlicna**
## PR contribution summary

**Why is this PR useful / good for?** *Please describe the problem(s) you're trying to address.*
* New `Prediction` object to store well-annotated predictions at inference time. 
* The dummy graphs append a more elaborate vector as the `GraphAttrs.NODE_PREDICTION`
* This used to be a single floating point number before (representing the probability of TP node / edge)
* This PR updates the optimiser & all helper functions to update dummy predictions accordingly
* Included 2 types of tests for sanity checks of prediction assignments & optimiser output

### List of proposed changes / linked issues & discussions

- [x] Resolves #9 
- [x] Resolves #191
- [x] Resolves #213 
- [x] Resolves #226
- [x] Resolves #228 

### What should a reviewer concentrate their feedback on?

+ 🏃 Scripts to check (optimiser + dummy prediction functions)
+ 🏃 Notebooks to run (especially `measure_metrics.ipynb`)
+ 💻 Code quality
+ 📝 Everything looks OK?

#### What type of PR is this? (check all applicable)

+ 🪄 Feature
+ 🧑‍💻 Code refactor / style

#### Added tests?

+ 🙋 yes, namely `test_dummy.py` & `test_optimiser.py`

---

## PR review summary

Describe what this PR does & **how you reviewed** the individual items, where needed:

* <TEXT>

*Some helper checks to tick off:*

* Focus on image annotation
* Focus on model training
* Could any optimization be applied?
* Is there any redundant code?
* Are there any spelling errors?

**In conclusion**, after my review, I'd like to:

* 🙋 ask some clarifying questions
* 🙅 suggest some specific changes

### CODE REVIEW

File: grace/base.py [hunk 1]
```diff
@@ -102,7 +102,7 @@ def delaunay_edges_from_nodes(
     # add edge nodes
     if update_graph:
         edge_attrs = {
-            GraphAttrs.EDGE_PREDICTION: 0.0,
+            GraphAttrs.EDGE_PREDICTION: (0, np.array([0.9, 0.1])),
```
Comment (quantumjot) [inline grace/base.py:105]: Can you elaborate on this - why are you using a tuple here? I assume this is `(label, logits)`?

In general, I'm not a huge fan of this as it involves indexing, and that can lead to unclear code and errors later. I would suggest that we have either a dictionary or a custom type, e.g.:

```python
class EdgePrediction:
    label: int
    logits: npt.NDArray
```

File: grace/evaluation/process.py [hunk 2]
```diff
@@ -28,30 +30,68 @@ def update_graph_with_dummy_predictions(
...
+        which are very close, whilst 0.1 will force larger separation between
+        them, i.e. TN between [0.0, 0.1) & TP between (0.9, 1.0].
+    - Only motifs with the same object identity should be linked (TODO)
     """
 
     # Make sure all objects are labelled according to object identity:
     nodes = list(G.nodes.data())
-    obj_identity = ["object_idx" in node for _, node in nodes]
-    assert all(obj_identity)
+    obj_identity = ["object_idx" in n and "label" in n for _, n in nodes]
```
Comment (quantumjot) [inline grace/evaluation/process.py:43]: Is `object_idx` guaranteed to be in the node?

File: grace/evaluation/process.py [hunk 3]
```diff
@@ -28,30 +30,68 @@ def update_graph_with_dummy_predictions(
...
     -----
     - Modifies the graph in place.
-    - Only motifs with the same object identity will be linked.
+    - The 'uncertainty' parameter separates the distributions of predictions
+        for the respective classes, e.g. 0.5 will produce a distribution of
+        attribute values for TN between [0.0, 0.5) & TP between (0.5, 1.0]
+        which are very close, whilst 0.1 will force larger separation between
+        them, i.e. TN between [0.0, 0.1) & TP between (0.9, 1.0].
+    - Only motifs with the same object identity should be linked (TODO)
     """
```
Comment (quantumjot) [inline grace/evaluation/process.py:40]: Seems like you should add a check that `node_uncertainty` and `edge_uncertainty` are in an acceptable range and raise an error if not.

File: grace/evaluation/process.py [hunk 4]
```diff
@@ -28,30 +30,68 @@ def update_graph_with_dummy_predictions(
...
-            node[GraphAttrs.NODE_PREDICTION] = pd
-        else:
-            node[GraphAttrs.NODE_PREDICTION] = 1 - pd
+        pd = np.random.random() * node_uncertainty
+
+        if all(obj_identity):
+            if node["label"] > 0:  # true positive node
+                node[GraphAttrs.NODE_PREDICTION] = (1, np.array([pd, 1 - pd]))
+            else:
+                node[GraphAttrs.NODE_PREDICTION] = (0, np.array([1 - pd, pd]))
```
Comment (quantumjot) [inline grace/evaluation/process.py:54]: I'm confused by what these are from this.

File: grace/evaluation/process.py [hunk 5]
```diff
@@ -28,30 +30,68 @@ def update_graph_with_dummy_predictions(
...
+            ):  # TP
+                edge[-1][GraphAttrs.EDGE_PREDICTION] = (
+                    1,
+                    np.array([pd, 1 - pd]),
+                )
+            else:
+                edge[-1][GraphAttrs.EDGE_PREDICTION] = (
+                    0,
+                    np.array([1 - pd, pd]),
+                )
```
Comment (quantumjot) [inline grace/evaluation/process.py:79]: Looking at the code further down (in the optimiser), I'm unsure why you need to store the probabilities as an array?

File: grace/base.py (hunk 1 above)
Comment (quantumjot) [reply grace/base.py:105]: What happens if you have multiple labels, i.e. the length of the array is variable?

File: grace/models/optimiser.py [hunk 6]
```diff
@@ -141,15 +141,19 @@ def optimise_graph(
...
-            Hypothesis(i=None, j=i, rho=n_dict[GraphAttrs.NODE_PREDICTION])
+            Hypothesis(
+                i=None, j=i, rho=n_dict[GraphAttrs.NODE_PREDICTION][1][0]
+            )
         )
     # build a set of link hypotheses
     for i, j, e_dict in graph.edges.data():
         hypotheses.append(
-            Hypothesis(i=i, j=j, rho=e_dict[GraphAttrs.EDGE_PREDICTION])
+            Hypothesis(i=i, j=j, rho=e_dict[GraphAttrs.EDGE_PREDICTION][1][1])
```
Comment (quantumjot) [inline grace/models/optimiser.py:156]: Again, do you need the array here? It seems like a simple probability would do

**Review by quantumjot — COMMENTED** [review]
Hi @KristinaUlicna - I've added a few comments, and I haven't tested the code

File: grace/base.py (hunk 1 above)
Comment (KristinaUlicna) [reply grace/base.py:105]: I primarily sources from the template code in `opt.ipynb` where this was the logic. @crangelsmith kindly suggested a dictionary, too, with `str` keys like "predicted_label", "probab_TN", "probab_TP" which is what the optimiser needs.

File: grace/evaluation/process.py (hunk 2 above)
Comment (KristinaUlicna) [reply grace/evaluation/process.py:43]: Not necessarily, the code chooses whether the GT labels come from the dummy or real annotations. That's what the assert `or` statement checks - some GT annotation must be present.

File: grace/evaluation/process.py (hunk 5 above)
Comment (KristinaUlicna) [reply grace/evaluation/process.py:79]: You don't, but that's how the label (=argmax) is calculated. It's stored for convenience and allows quick checks if the sum adds up to 1.

File: grace/models/optimiser.py (hunk 6 above)
Comment (KristinaUlicna) [reply grace/models/optimiser.py:156]: Nodes and edges need probability of a different class for the optimiser to work, as stated in the documentation.

Comment (quantumjot) [reply grace/models/optimiser.py:156]: Sure, but you're indexing by that anyway with the `GraphAttrs` key. I think you've clarified that what you want to store is the probability distribution over the labels - which seems to make sense. I would just make that clear in the documentation and change how you store it in the edge attributes, as noted above

**Review by mooniean — COMMENTED** [review]
The scripts were running and the notebooks ran fully.

I noticed that the simulated graphs make curves outside of the random dots, is this intended? 
image

I can add as a feature/issue to fix if not intended!
Otherwise looks good

**Comment by KristinaUlicna** [comment]
The shapes overarching the random graph are mentioned in issue #173 - will be fixed at some other time.

File: grace/models/optimiser.py (hunk 6 above)
Comment (KristinaUlicna) [reply grace/models/optimiser.py]: All implemented in the latest update - thanks for the feedback & helpful suggestions on how to improve the code!

File: grace/evaluation/process.py (hunk 3 above)
Comment (KristinaUlicna) [reply grace/evaluation/process.py:52]: Now implemented!

**Review by marjanfamili — APPROVED** [review]
I have tested the notebook for higher number of nodes and motifs (50,100,150).

The implementation looks neat. 

All looks good

File: grace/base.py [hunk 7]
```diff
@@ -34,6 +34,15 @@ class Annotation(enum.IntEnum):
     UNKNOWN = 2
 
 
+class Prediction(enum.Enum):
+    """Values of predicted label & logits."""
+
+    LABEL: int = "predicted class label (argmax)"
+    PROB_TN: float = "probability of true negative detection (softmax)"
+    PROB_TP: float = "probability of true positive detection (softmax)"
+    # PROB_UNKNOWN: float = "probability of unknown detection (softmax)"
```
Comment (quantumjot) [inline grace/base.py:43]: I don't have time to review this, but this doesn't look correct to me. It shouldn't be an enum and the typing is wrong given the values. Aso, the logits are unscaled, so perhaps change the docstring

Comment (quantumjot) [reply grace/base.py:43]: I think it should just be a regular class, or a dataclass

Comment (KristinaUlicna) [reply grace/base.py:43]: I don't think I fully understand what's the difference between this and the `GraphAttrs` class as defined above, which also takes in a mixture of datatypes, but I'm following the suggestion & implementing the change now.

File: grace/base.py [hunk 8]
```diff
@@ -37,27 +37,48 @@ class Annotation(enum.IntEnum):
...
-    PROB_TN : float
-        Probability of true negative detection (softmax)
-    PROB_TP : float
-        Probability of true positive detection (softmax)
+    softmax_probabs : npt.NDArray
+        Array or normalised softmax probs as predicted by classifier.
+
+    Methods
+    -------
+    label : self -> Annotation
```
Comment (quantumjot) [inline grace/base.py:49]:
```suggestion
    label : Annotation
```

similar for the others below

File: grace/base.py [hunk 9]
```diff
@@ -37,27 +37,48 @@ class Annotation(enum.IntEnum):
...
 
     Parameters
     ----------
-    LABEL : int
-        Predicted class label (argmax)
-    PROB_TN : float
-        Probability of true negative detection (softmax)
-    PROB_TP : float
-        Probability of true positive detection (softmax)
+    softmax_probabs : npt.NDArray
```
Comment (quantumjot) [inline grace/base.py:44]: I would either use `softmax_probabilities` or `softmax_probs`

File: grace/base.py [hunk 10]
```diff
@@ -37,27 +37,48 @@ class Annotation(enum.IntEnum):
...
     """
 
-    LABEL: int
-    PROB_TN: float
-    PROB_TP: float
+    softmax_probabs: npt.NDArray
+
+    def __post_init__(self):
+        assert self.softmax_probabs.ndim == 1
+        assert self.softmax_probabs.shape[0] >= 2
```
Comment (quantumjot) [inline grace/base.py:66]:
```suggestion
        assert self.softmax_probabs.shape[0] == len(Annotation)
```

I would make this explicit, the array needs to be the same as the number of possible labels

File: grace/evaluation/process.py [hunk 11]
```diff
@@ -63,33 +63,17 @@ def update_graph_with_dummy_predictions(
         if all(obj_identity):
             # true positive node:
             if node["label"] > 0:
-                pred = {
-                    Prediction.LABEL: 1,
-                    Prediction.PROB_TN: pd,
-                    Prediction.PROB_TP: 1 - pd,
-                }
+                pred = Prediction(np.array([pd, 1 - pd]))
```
Comment (quantumjot) [inline grace/evaluation/process.py:66]: so, then you can add probability unknown of 0.0 here:

```suggestion
                pred = Prediction(np.array([pd, 1 - pd, 0]))
```

File: grace/evaluation/process.py [hunk 12]
```diff
@@ -215,11 +183,11 @@ def assume_annotations_from_dummy_predictions(G: nx.Graph) -> None:
     """Translates between dummy predictions & GT labels."""
 
     for _, node in G.nodes(data=True):
-        pred_label = node[GraphAttrs.NODE_PREDICTION][Prediction.LABEL]
+        pred_label = node[GraphAttrs.NODE_PREDICTION].label()
```
Comment (quantumjot) [inline grace/evaluation/process.py:186]: No need to call this
```suggestion
        pred_label = node[GraphAttrs.NODE_PREDICTION].label
```

File: grace/evaluation/process.py [hunk 13]
```diff
@@ -215,11 +183,11 @@ def assume_annotations_from_dummy_predictions(G: nx.Graph) -> None:
...
         node[GraphAttrs.NODE_GROUND_TRUTH] = Annotation(pred_label)
 
     for _, _, edge in G.edges(data=True):
-        pred_label = edge[GraphAttrs.EDGE_PREDICTION][Prediction.LABEL]
+        pred_label = node[GraphAttrs.NODE_PREDICTION].label()
```
Comment (quantumjot) [inline grace/evaluation/process.py:190]: same here

**Review by quantumjot — APPROVED** [review]
Looks much improved! A few final things to consider.

File: grace/base.py (hunk 10 above)
Comment (KristinaUlicna) [reply grace/base.py:66]: With the current implementation of the classifier, which outputs 2 labels and ignores `UNKNOWN`, perhaps it's better to check for `len(Annotation)` or `len(Annotation) - 1`...?

File: grace/evaluation/process.py (hunk 12 above)
Comment (KristinaUlicna) [reply grace/evaluation/process.py:186]: Sorry, my oversight!

File: grace/evaluation/process.py (hunk 11 above)
Comment (KristinaUlicna) [reply grace/evaluation/process.py:66]: Ohhh ok, I see what you mean. Just assign the vector which equals the length of the annotation, even though we're not learning it. Implementing now.

File: grace/base.py [hunk 14]
```diff
@@ -102,16 +149,16 @@ def delaunay_edges_from_nodes(
     # add edge nodes
     if update_graph:
         edge_attrs = {
-            GraphAttrs.EDGE_PREDICTION: 0.0,
+            # GraphAttrs.EDGE_PREDICTION: (0, np.array([0.9, 0.1])),
```
Comment (quantumjot) [inline grace/base.py:152]: Should this be uncommented too?

Comment (KristinaUlicna) [reply grace/base.py:152]: It doesn't have to be there as the `update_graph` only needs to construct the edges with UNKNOWN GT labels. Predictions can / will be added afterwards, if needed, either with the update dummy function or from real classifier.

Comment (KristinaUlicna) [reply grace/base.py]: I took it out completely.
